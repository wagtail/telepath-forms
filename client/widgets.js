import { runInlineScripts } from './utils';


export const adapters = {};


export class BoundWidget {
  constructor(
    elementOrNodeList,
    name,
    idForLabel,
  ) {
    // if elementOrNodeList not iterable, it must be a single element
    const nodeList = elementOrNodeList.forEach
      ? elementOrNodeList
      : [elementOrNodeList];

    // look for an input element with the given name, as either a direct element of nodeList
    // or a descendant
    const selector = `:is(input,select,textarea,button)[name="${name}"]`;

    for (let i = 0; i < nodeList.length; i += 1) {
      const element = nodeList[i];
      if (element.nodeType === Node.ELEMENT_NODE) {
        if (element.matches(selector)) {
          this.input = element;
          break;
        } else {
          const input = element.querySelector(selector);
          if (input) {
            this.input = input;
            break;
          }
        }
      }
    }

    this.idForLabel = idForLabel;
  }

  getValue() {
    return this.input.value;
  }

  getState() {
    return this.input.value;
  }

  setState(state) {
    this.input.value = state;
  }

  getTextLabel(opts) {
    const val = this.getValue();
    if (typeof val !== 'string') return null;
    const maxLength = opts && opts.maxLength;
    if (maxLength && val.length > maxLength) {
      return val.substring(0, maxLength - 1) + 'â€¦';
    }
    return val;
  }

  focus() {
    this.input.focus();
  }
}

export class Widget {
  constructor(html, idPattern) {
    this.html = html;
    this.idPattern = idPattern;
  }

  boundWidgetClass = BoundWidget;

  render(
    placeholder,
    name,
    id,
    initialState,
    options = {},
  ) {
    const html = this.html.replace(/__NAME__/g, name).replace(/__ID__/g, id);
    const idForLabel = this.idPattern.replace(/__ID__/g, id);

    /* write the HTML into a temp container to parse it into a node list */
    const tempContainer = document.createElement('div');
    tempContainer.innerHTML = html.trim();
    const childNodes = Array.from(tempContainer.childNodes);

    /* replace the placeholder with the new nodes */
    placeholder.replaceWith(...childNodes);

    const childElements = childNodes.filter(
      (node) => node.nodeType === Node.ELEMENT_NODE,
    );

    /* execute any scripts in the new element(s) */
    childElements.forEach((element) => {
      runInlineScripts(element);
    });

    // Add any extra attributes we received to the first element of the widget
    if (typeof options?.attributes === 'object') {
      Object.entries(options.attributes).forEach(([key, value]) => {
        childElements[0].setAttribute(key, value);
      });
    }

    // eslint-disable-next-line new-cap
    const boundWidget = new this.boundWidgetClass(
      childElements.length === 1 ? childElements[0] : childNodes,
      name,
      idForLabel,
    );
    boundWidget.setState(initialState);
    return boundWidget;
  }
}
adapters['telepath.forms.Widget'] = Widget;


export class BoundCheckboxInput extends BoundWidget {
  getValue() {
    return this.input.checked;
  }

  getState() {
    return this.input.checked;
  }

  setState(state) {
    this.input.checked = state;
  }
}

export class CheckboxInput extends Widget {
  boundWidgetClass = BoundCheckboxInput;
}
adapters['telepath.forms.CheckboxInput'] = CheckboxInput;


export class BoundRadioSelect {
  constructor(element, name, idForLabel) {
    this.element = element;
    this.name = name;
    this.idForLabel = idForLabel;
    this.isMultiple = !!this.element.querySelector(
      `input[name="${name}"][type="checkbox"]`,
    );
    this.selector = `input[name="${name}"]:checked`;
  }

  getValue() {
    if (this.isMultiple) {
      return Array.from(this.element.querySelectorAll(this.selector)).map(
        (el) => el.value,
      );
    }
    return this.element.querySelector(this.selector)?.value;
  }

  getState() {
    return Array.from(this.element.querySelectorAll(this.selector)).map(
      (el) => el.value,
    );
  }

  setState(state) {
    const inputs = this.element.querySelectorAll(`input[name="${this.name}"]`);
    for (let i = 0; i < inputs.length; i += 1) {
      inputs[i].checked = state.includes(inputs[i].value);
    }
  }

  focus() {
    this.element.querySelector(`input[name="${this.name}"]`)?.focus();
  }
}

export class RadioSelect extends Widget {
  boundWidgetClass = BoundRadioSelect;
}
adapters['telepath.forms.RadioSelect'] = RadioSelect;


export class BoundSelect extends BoundWidget {
  getTextLabel() {
    return Array.from(this.input.selectedOptions)
      .map((option) => option.text)
      .join(', ');
  }

  getValue() {
    if (this.input.multiple) {
      return Array.from(this.input.selectedOptions).map(
        (option) => option.value,
      );
    }
    return this.input.value;
  }

  getState() {
    return Array.from(this.input.selectedOptions).map((option) => option.value);
  }

  setState(state) {
    const options = this.input.options;
    for (let i = 0; i < options.length; i += 1) {
      options[i].selected = state.includes(options[i].value);
    }
  }
}

export class Select extends Widget {
  boundWidgetClass = BoundSelect;
}
adapters['telepath.forms.Select'] = Select;


export class ValidationError {
  constructor(messages) {
    this.messages = messages;
  }
}
adapters['telepath.forms.ValidationError'] = ValidationError;
