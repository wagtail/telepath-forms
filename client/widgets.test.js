import { Widget, CheckboxInput, RadioSelect, Select } from "./widgets";


beforeEach(() => {
  document.body.innerHTML = "";
});


describe('Widget', () => {
  it('can be rendered', () => {
    const widget = new Widget(
      '<input type="text" name="__NAME__" id="__ID__">', "__ID__"
    );

    const placeholder = document.createElement('div');
    document.body.appendChild(placeholder);

    const boundWidget = widget.render(placeholder, "name", "id_name", "Bob", {
      attributes: {"aria-describedby": "id_name-help-text"}
    });

    expect(document.body.innerHTML).toMatchSnapshot();
    expect(boundWidget.idForLabel).toBe("id_name");
    expect(boundWidget.getValue()).toBe("Bob");
    expect(boundWidget.getState()).toBe("Bob");

    boundWidget.setState("Alice");
    expect(boundWidget.getValue()).toBe("Alice");
    expect(boundWidget.getState()).toBe("Alice");
    expect(boundWidget.getTextLabel()).toBe("Alice");
    expect(boundWidget.getTextLabel({maxLength: 4})).toBe("Aliâ€¦");

    boundWidget.focus();
    expect(document.activeElement).toBe(boundWidget.input);
  });

  it('can be rendered when input is not a top-level element', () => {
    const widget = new Widget(
      '<div><input type="text" name="__NAME__" id="__ID__"></div>', "__ID__"
    );

    const placeholder = document.createElement('div');
    document.body.appendChild(placeholder);

    const boundWidget = widget.render(placeholder, "name", "id_name", "Bob", {
      attributes: {"aria-describedby": "id_name-help-text"}
    });

    expect(document.body.innerHTML).toMatchSnapshot();
    expect(boundWidget.getValue()).toBe("Bob");
    expect(boundWidget.getState()).toBe("Bob");
    boundWidget.setState("Alice");
    expect(boundWidget.getValue()).toBe("Alice");
    expect(boundWidget.getState()).toBe("Alice");
  });
});

describe('CheckboxInput', () => {
  it('can be rendered', () => {
    const widget = new CheckboxInput(
      '<input type="checkbox" name="__NAME__" id="__ID__">', "__ID__"
    );

    const placeholder = document.createElement('div');
    document.body.appendChild(placeholder);

    const boundWidget = widget.render(placeholder, "extra_cheese", "id_extra_cheese", false, {
      attributes: {"aria-describedby": "id_extra_cheese-help-text"}
    });

    expect(document.body.innerHTML).toMatchSnapshot();
    expect(boundWidget.getValue()).toBe(false);
    expect(boundWidget.getState()).toBe(false);
    boundWidget.setState(true);
    expect(boundWidget.getValue()).toBe(true);
    expect(boundWidget.getState()).toBe(true);
  });
});

describe('RadioSelect', () => {
  it('can be rendered', () => {
    const widget = new RadioSelect(`
      <div id="__ID__">
        <div>
          <label for="__ID___0">
            <input type="radio" name="__NAME__" value="red" id="__ID___0">
            Red
          </label>
        </div>
        <div>
          <label for="__ID___1">
            <input type="radio" name="__NAME__" value="green" id="__ID___1">
            Green
          </label>
        </div>
        <div>
          <label for="__ID___2">
            <input type="radio" name="__NAME__" value="blue" id="__ID___2">
            Blue
          </label>
        </div>
      </div>
    `, "");

    const placeholder = document.createElement('div');
    document.body.appendChild(placeholder);

    const boundWidget = widget.render(placeholder, "color", "id_color", ["green"], {
      attributes: {"aria-describedby": "id_color-help-text"}
    });

    expect(document.body.innerHTML).toMatchSnapshot();
    expect(boundWidget.idForLabel).toBe("");
    expect(boundWidget.getValue()).toBe("green");
    expect(boundWidget.getState()).toStrictEqual(["green"]);

    boundWidget.setState(["blue"]);
    expect(boundWidget.getValue()).toBe("blue");
    expect(boundWidget.getState()).toStrictEqual(["blue"]);

    boundWidget.focus();
    expect(document.activeElement.value).toBe("red");
  });
});

describe('multiple choice RadioSelect', () => {
  it('can be rendered', () => {
    const widget = new RadioSelect(`
      <div id="__ID__">
        <div>
          <label for="__ID___0">
            <input type="checkbox" name="__NAME__" value="red" id="__ID___0">
            Red
          </label>
        </div>
        <div>
          <label for="__ID___1">
            <input type="checkbox" name="__NAME__" value="green" id="__ID___1">
            Green
          </label>
        </div>
        <div>
          <label for="__ID___2">
            <input type="checkbox" name="__NAME__" value="blue" id="__ID___2">
            Blue
          </label>
        </div>
      </div>
    `, "");

    const placeholder = document.createElement('div');
    document.body.appendChild(placeholder);

    const boundWidget = widget.render(placeholder, "color", "id_color", ["red", "green"], {
      attributes: {"aria-describedby": "id_color-help-text"}
    });

    expect(document.body.innerHTML).toMatchSnapshot();
    expect(boundWidget.getValue()).toStrictEqual(["red", "green"]);
    expect(boundWidget.getState()).toStrictEqual(["red", "green"]);
    boundWidget.setState(["green", "blue"]);
    expect(boundWidget.getValue()).toStrictEqual(["green", "blue"]);
    expect(boundWidget.getState()).toStrictEqual(["green", "blue"]);
  });
});

describe('Select', () => {
  it('can be rendered', () => {
    const widget = new Select(`
      <select name="__NAME__" id="__ID__">
        <option value="red">Red</option>
        <option value="green">Green</option>
        <option value="blue">Blue</option>
      </select>
    `, "__ID__");

    const placeholder = document.createElement('div');
    document.body.appendChild(placeholder);

    const boundWidget = widget.render(placeholder, "color", "id_color", ["green"], {
      attributes: {"aria-describedby": "id_color-help-text"}
    });

    expect(document.body.innerHTML).toMatchSnapshot();
    expect(boundWidget.getValue()).toBe("green");
    expect(boundWidget.getState()).toStrictEqual(["green"]);
    expect(boundWidget.getTextLabel()).toBe("Green");

    boundWidget.setState(["blue"]);
    expect(boundWidget.getValue()).toBe("blue");
    expect(boundWidget.getState()).toStrictEqual(["blue"]);
  });
});

describe('multiple choice Select', () => {
  it('can be rendered', () => {
    const widget = new Select(`
      <select name="__NAME__" id="__ID__" multiple>
        <option value="red">Red</option>
        <option value="green">Green</option>
        <option value="blue">Blue</option>
      </select>
    `, "__ID__");

    const placeholder = document.createElement('div');
    document.body.appendChild(placeholder);

    const boundWidget = widget.render(placeholder, "color", "id_color", ["green"], {
      attributes: {"aria-describedby": "id_color-help-text"}
    });

    expect(document.body.innerHTML).toMatchSnapshot();
    expect(boundWidget.getValue()).toStrictEqual(["green"]);
    expect(boundWidget.getState()).toStrictEqual(["green"]);
    expect(boundWidget.getTextLabel()).toBe("Green");

    boundWidget.setState(["blue"]);
    expect(boundWidget.getValue()).toStrictEqual(["blue"]);
    expect(boundWidget.getState()).toStrictEqual(["blue"]);
  });
});
