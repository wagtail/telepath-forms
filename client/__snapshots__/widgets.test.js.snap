// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxInput can be rendered 1`] = `"<input type="checkbox" name="extra_cheese" id="id_extra_cheese" aria-describedby="id_extra_cheese-help-text">"`;

exports[`RadioSelect can be rendered 1`] = `
"<div id="id_color" aria-describedby="id_color-help-text">
        <div>
          <label for="id_color_0">
            <input type="radio" name="color" value="red" id="id_color_0">
            Red
          </label>
        </div>
        <div>
          <label for="id_color_1">
            <input type="radio" name="color" value="green" id="id_color_1">
            Green
          </label>
        </div>
        <div>
          <label for="id_color_2">
            <input type="radio" name="color" value="blue" id="id_color_2">
            Blue
          </label>
        </div>
      </div>"
`;

exports[`Select can be rendered 1`] = `
"<select name="color" id="id_color" aria-describedby="id_color-help-text">
        <option value="red">Red</option>
        <option value="green">Green</option>
        <option value="blue">Blue</option>
      </select>"
`;

exports[`Widget can be rendered 1`] = `"<input type="text" name="name" id="id_name" aria-describedby="id_name-help-text">"`;

exports[`Widget can be rendered when input is not a top-level element 1`] = `"<div aria-describedby="id_name-help-text"><input type="text" name="name" id="id_name"></div>"`;

exports[`multiple choice RadioSelect can be rendered 1`] = `
"<div id="id_color" aria-describedby="id_color-help-text">
        <div>
          <label for="id_color_0">
            <input type="checkbox" name="color" value="red" id="id_color_0">
            Red
          </label>
        </div>
        <div>
          <label for="id_color_1">
            <input type="checkbox" name="color" value="green" id="id_color_1">
            Green
          </label>
        </div>
        <div>
          <label for="id_color_2">
            <input type="checkbox" name="color" value="blue" id="id_color_2">
            Blue
          </label>
        </div>
      </div>"
`;

exports[`multiple choice Select can be rendered 1`] = `
"<select name="color" id="id_color" multiple="" aria-describedby="id_color-help-text">
        <option value="red">Red</option>
        <option value="green">Green</option>
        <option value="blue">Blue</option>
      </select>"
`;
